### PR TITLE
Bogdan

### PR DESCRIPTION
**Pull request by bogdanishere**
Am adăugat librăria React Query pentru optimizarea gestionării datelor pe ruta /admin/quickround
- pentru a beneficia de refetch automat când o întrebare este ștearsă sau editată.
- oferă caching, sincronizarea datelor în fundal și actualizarea automată a datelor pe UI.
- îmbunătățește performanța și gestionarea stării în aplicație, fără a mai fi nevoie de a scrie logica manuală pentru refetch.

Am creat un folder "network" care centralizează toate tipurile de request-uri, deoarece această structură face codul mai lizibil și situațional dacă cererile sunt reutilizate.
Am adăugat un custom hook cu React Query pentru a simplifica logica și îmbunătăți claritatea codului.
În app.tsx am adăugat:
- Suspense pentru a afișa un spinner atunci când pagina se încarcă
- Integrat QueryClient pentru a permite utilizarea librăriei React Query în aplicație.

### CODE REVIEW

File: client/package.json [hunk 1]
```diff
@@ -10,8 +10,11 @@
     "preview": "vite preview"
   },
   "dependencies": {
+    "@tanstack/react-query": "^5.54.1",
+    "client": "file:",
```
Comment (cristim67) [inline client/package.json:14]: for what you use this dependency? `"client": "file:"`

File: client/src/App.tsx [hunk 2]
```diff
@@ -8,28 +8,34 @@ import Ranking from "./views/Ranking";
...
-          <Route path="/ranking" element={<Ranking />} />
-          <Route path="/admin/addTeam" element={<AdminAddTeam />} />
-          <Route path="/admin/quickround" element={<AdminQuickRound />} />
-          <Route
-            path="/admin/addquestionmainround"
-            element={<AdminAddQuestionMainRound />}
-          />
-          <Route path="*" element={<Navigate to="/login" />} />
-        </Routes>
+        <Suspense fallback={<Spinner />}>
```
Comment (cristim67) [inline client/src/App.tsx]: you have a `Spinner` for fallback, but no error boundary.
 
 should be like this: 
 
 ```
<ErrorBoundary fallback={<ErrorComponent />}>
  <Suspense fallback={<Spinner />}>
    {/* Routes go here */}
  </Suspense>
</ErrorBoundary>

```

File: client/src/components/AddQuestionForm.tsx [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    const sendObject = {
+      question: filteredData.question,
+      answers: answersToSend,
+    };
+
+    try {
+      const response = await adminQuickRound.createQuestion(sendObject);
+
+      console.log("Question created successfully:", response);
+      queryClient.invalidateQueries({ queryKey: ["questions"] });
```
Comment (cristim67) [inline client/src/components/AddQuestionForm.tsx:83]: i think here, you need to put an `await` keyword

File: client/src/components/AddQuestionForm.tsx [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  const {
+    register,
+    handleSubmit,
+    formState: { errors, isSubmitting },
+  } = useForm<FormValues>();
+
+  const [numOfAnswers, setNumOfAnswers] = useState(3);
+  const [correctScore, setCorrectScore] = useState<boolean>(true);
+
+  const handleAnswerChange = (event: React.ChangeEvent<HTMLInputElement>) => {
```
Comment (cristim67) [inline client/src/components/AddQuestionForm.tsx]: import `React` if you use it.

File: client/src/components/AddQuestionForm.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+    const sendObject = {
+      question: filteredData.question,
+      answers: answersToSend,
+    };
+
+    try {
+      const response = await adminQuickRound.createQuestion(sendObject);
+
+      console.log("Question created successfully:", response);
```
Comment (cristim67) [inline client/src/components/AddQuestionForm.tsx]: remove the console.log

File: client/src/App.tsx [hunk 6]
```diff
@@ -8,28 +8,34 @@ import Ranking from "./views/Ranking";
...
+          <Routes>
+            <Route path="/login" element={<Login />} />
+            <Route path="/register" element={<Register />} />
+            <Route path="/mainround" element={<MainRound />} />
+            <Route path="/quickround" element={<QuickRound />} />
+            <Route path="/scoredisplay" element={<ScoreDisplay />} />
+            <Route path="/ranking" element={<Ranking />} />
+            <Route path="/admin/addTeam" element={<AdminAddTeam />} />
+            <Route path="/admin/quickround" element={<AdminQuickRound />} />
```
Comment (cristim67) [inline client/src/App.tsx]: i think you need an admin wrapper for `/admin/*` route wildcard to check if the user is admin.

File: client/src/components/AddQuestionForm.tsx [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+  closeModal: () => void;
+}) {
+  const queryClient = useQueryClient();
+  const {
+    register,
+    handleSubmit,
+    formState: { errors, isSubmitting },
+  } = useForm<FormValues>();
+
+  const [numOfAnswers, setNumOfAnswers] = useState(3);
```
Comment (cristim67) [inline client/src/components/AddQuestionForm.tsx]: default: 6

File: client/src/components/AddQuestionForm.tsx [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+          ))}
+
+          <Button type="submit" addNewContent="w-full mt-4">
+            Submit
+          </Button>
+        </form>
+      </div>
+    </div>
+  );
+}
```
Comment (cristim67) [inline client/src/components/AddQuestionForm.tsx]: there’s no feedback for users if they input invalid data. display validation error messages next to the corresponding fields. should be like this: 

```
{errors.answers?.[index]?.numberOfResponse && (
  <p className="text-red-500">Number of points is required.</p>
)}
```

File: client/src/components/AddQuestionForm.tsx [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    if (totalResponses !== 100) {
+      console.error("Total responses must be 100");
+      setCorrectScore(false);
+      return;
+    }
+
+    const answersToSend: AnswerSendProp = {};
+    filteredData.answers.forEach((answer) => {
+      answersToSend[answer.answer] = +answer.numberOfResponse;
+      console.log("answersToSend", answersToSend);
```
Comment (cristim67) [inline client/src/components/AddQuestionForm.tsx]: remove this console.log

File: client/src/components/AnswersAndVoters.tsx [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  [key: string]: number;
+}
+
+export default function AnswersAndVotes({ answers }: { answers: Answer }) {
+  return (
+    <div className="space-y-2 mb-4">
+      {Object.entries(answers).map(([answer, count]) => (
+        <div key={answer} className="flex justify-between">
+          <span className="text-gray-700">{answer}</span>
+          <span className="font-medium text-gray-900">Votes: {count}</span>
```
Comment (cristim67) [inline client/src/components/AnswersAndVoters.tsx]: why you name, `votes` everywhere? there are points

File: client/src/components/Button.tsx [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+/**
+ *
+ * @param children The button content.
+ * @param addNewContent A function to add new styles to the button. If you use `className`, it will overwrite all the styles.
+ * @returns A button with custom styles or default styles.
+ */
+
+export default function Button({
```
Comment (cristim67) [inline client/src/components/Button.tsx]: this was necessary?

File: client/src/components/AddQuestionForm.tsx (hunk 9 above)
Comment (bogdanishere) [reply client/src/components/AddQuestionForm.tsx]: do i need to remove console.error as well?

Comment (cristim67) [reply client/src/components/AddQuestionForm.tsx]: if the backend returns something sensitive, yes.

File: client/src/components/Button.tsx (hunk 11 above)
Comment (bogdanishere) [reply client/src/components/Button.tsx]: Yes, to have a custom button so we don't need to write the code over and over again for every button on the page, and the comments above are to make it clear about the button.

File: client/src/components/AnswersAndVoters.tsx (hunk 10 above)
Comment (bogdanishere) [reply client/src/components/AnswersAndVoters.tsx]: I will change that right now

File: client/src/components/EditQuestionForm.tsx [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+  question: string;
+  answers: { answer: string; numberOfResponse: number }[];
+}
+
+export default function EditQuestionForm({
+  questionData,
+  onSubmit,
+  onClose,
+}: EditQuestionFormProps) {
+  const initialAnswers = Object.entries(questionData.answers).map(
```
Comment (cristim67) [inline client/src/components/EditQuestionForm.tsx]: Use `useMemo` to optimize `initialAnswers`.

```
const initialAnswers = useMemo(
  () =>
    Object.entries(questionData.answers).map(([answer, numberOfResponse]) => ({
      answer,
      numberOfResponse,
    })),
  [questionData.answers]
);

```

File: client/src/components/EditQuestionForm.tsx [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    handleSubmit,
+    formState: { errors },
+  } = useForm<FormValues>({
+    defaultValues: {
+      question: questionData.question,
+      answers: initialAnswers,
+    },
+  });
+
+  const handleFormSubmit: SubmitHandler<FormValues> = (data) => {
```
Comment (cristim67) [inline client/src/components/EditQuestionForm.tsx:52]: Add stricter validation for fields, such as ensuring positive numbers for points.

File: client/src/components/EditQuestionForm.tsx [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+  question: string;
+  answers: { [key: string]: number };
+}
+
+interface FormValues {
+  question: string;
+  answers: { answer: string; numberOfResponse: number }[];
+}
+
+export default function EditQuestionForm({
```
Comment (cristim67) [inline client/src/components/EditQuestionForm.tsx:26]: Add custom error messages for better user feedback.

File: server/package.json [hunk 15]
```diff
@@ -21,6 +21,7 @@
     "pg": "^8.12.0",
     "pg-hstore": "^2.3.4",
     "sequelize": "^6.37.3",
+    "server": "file:",
```
Comment (cristim67) [inline server/package.json:24]: for what you use this dependency? "server": "file:"

File: client/src/components/AddQuestionForm.tsx (hunk 8 above)
Comment (bogdanishere) [reply client/src/components/AddQuestionForm.tsx]: The error prevents the user/admin from submitting the number of points if it is not specified in the form.

File: client/src/components/FormInputField.tsx [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+          {...props}
+          className={`block w-full max-w-md bg-transparent outline-none border-b-2 py-2 px-4 ${
+            error
+              ? "text-red-500 border-red-600"
+              : "text-gray-900 border-purple-900"
+          }`}
+        />
+        {inputGroupElement}
+      </div>
+      {error && <p className="text-sm text-red-500 mt-1">{error.message}</p>}
```
Comment (cristim67) [inline client/src/components/FormInputField.tsx]: make it more robust, you might want to ensure that error.message exists before rendering:

`{error?.message && <p className="text-sm text-red-500 mt-1">{error.message}</p>}`

File: client/src/components/Modal.tsx [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    document.addEventListener("mousedown", handleClickOutside);
+    return () => {
+      document.removeEventListener("mousedown", handleClickOutside);
+    };
+  }, [ref, handler]);
+}
+
+const ModalContext = createContext<ModalContextProps | undefined>(undefined);
+
+export function Modal({ children, isOpen, onClose }: ModalProps) {
```
Comment (cristim67) [inline client/src/components/Modal.tsx:61]: It’s good practice to support keyboard users by adding the ability to close the modal with the `Escape` key. This can be done by adding a keydown listener:

```
useEffect(() => {
  function handleKeyDown(event: KeyboardEvent) {
    if (event.key === "Escape") {
      onClose();
    }
  }
  document.addEventListener("keydown", handleKeyDown);
  return () => {
    document.removeEventListener("keydown", handleKeyDown);
  };
}, [onClose]);
```

File: client/src/components/QuestionBox.tsx [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+
+  interface UpdatedDataProps {
+    question: string;
+    answers: { [key: string]: number };
+  }
+
+  const handleEditSubmit = async (data: UpdatedDataProps) => {
+    try {
+      await adminQuickRound.updateQuestion(question.id, data);
+      queryClient.invalidateQueries({ queryKey: ["questions"] });
```
Comment (cristim67) [inline client/src/components/QuestionBox.tsx:52]: i think here you need an `await` keyword

File: client/src/components/QuestionBox.tsx [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+  return new Date(date).toLocaleString();
+}
+
+export default function QuestionBox({ question }: { question: QuestionProps }) {
+  const queryClient = useQueryClient();
+  const [isModalOpen, setIsModalOpen] = useState(false);
+  async function handleDelete(id: number) {
+    try {
+      await adminQuickRound.deleteQuestion(id);
+      queryClient.invalidateQueries({ queryKey: ["questions"] });
```
Comment (cristim67) [inline client/src/components/QuestionBox.tsx:34]: i think here you need an `await` keyword

File: client/src/newtork/adminQuickRound.ts [hunk 20]
```diff
@@ -0,0 +1,61 @@
+const url = import.meta.env.VITE_API_URL;
+const token = import.meta.env.VITE_JWT_TOKEN;
+
+export const getQuestions = async () => {
```
Comment (cristim67) [inline client/src/newtork/adminQuickRound.ts:4]: add error handling and a try catch block.

File: client/src/components/QuestionBox.tsx (hunk 18 above)
Comment (bogdanishere) [reply client/src/components/QuestionBox.tsx:52]: `queryClient.invalidateQueries({ queryKey: ["questions"] });` is a non-blocking operation, and the UI and application logic don't need to wait for this process to complete.

File: client/src/components/QuestionBox.tsx (hunk 19 above)
Comment (bogdanishere) [reply client/src/components/QuestionBox.tsx:34]: `queryClient.invalidateQueries({ queryKey: ["questions"] });` is a non-blocking operation, and the UI and application logic don't need to wait for this process to complete.

File: client/src/App.tsx [hunk 21]
```diff
@@ -8,28 +8,34 @@ import Ranking from "./views/Ranking";
...
+            <Route
+              path="/admin/addquestionmainround"
+              element={<AdminAddQuestionMainRound />}
+            />
+            <Route path="*" element={<Navigate to="/login" />} />
```
Comment (MDGcode) [inline client/src/App.tsx]: If you are on a inexistent page, you no longer get defaulted to login.
Please fix this issue

File: client/src/views/PageNotFetch.tsx
Comment (MDGcode) [inline client/src/views/PageNotFetch.tsx:1]: Why did you add this feature.
Also I have a hard time getting it to work locally.

File: client/src/views/AdminQuickRound.tsx
Comment (MDGcode) [inline client/src/views/AdminQuickRound.tsx:1]: You were supposed to make the feature for the /admin/addquestionmainround not /admin/addquestionmainround

The quick Round feature is allocated to another person

See the documentation

File: client/src/components/AddQuestionForm.tsx [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+        setNumOfAnswers(value);
+      }
+    }
+  };
+
+  interface AnswerSendProp {
+    [key: string]: number;
+  }
+
+  const onSubmit: SubmitHandler<FormValues> = async (data) => {
```
Comment (MDGcode) [inline client/src/components/AddQuestionForm.tsx:51]: Also check for answers with 0 score

File: client/src/components/QuestionBox.tsx [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+          </h2>
+          <AnswersAndVotes answers={question.answers} />
+          <div className="text-xs text-gray-700">
+            Created At: {convertDate(question.createdAt)}
+          </div>
+          <div className="text-xs text-gray-700">
+            Updated At: {convertDate(question.updatedAt)}
+          </div>
+        </div>
+        <div className="flex flex-col justify-center items-center gap-5">
```
Comment (MDGcode) [inline client/src/components/QuestionBox.tsx:73]: Replace delete and edit text with icons ( use react icons library).

Find a suitable icons and place them in the top right corner

File: client/src/components/QuestionBox.tsx [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+      queryClient.invalidateQueries({ queryKey: ["questions"] });
+      setIsModalOpen(false);
+    } catch (error) {
+      console.error("Failed to update question", error);
+    }
+  };
+
+  return (
+    <>
+      <div className="border border-gray-400 p-8 rounded-lg shadow-[0_0_10px_rgba(0,0,0,0.3)] bg-indigo-300 w-full grid grid-cols-2 gap-20">
```
Comment (MDGcode) [inline client/src/components/QuestionBox.tsx]: The question boxes are too large, keep in mind that we may have up to 100 questions

File: client/src/views/AdminQuickRound.tsx [hunk 25]
```diff
@@ -1,7 +1,64 @@
...
+  if (isError) return <PageNotFetch />;
+
   return (
-    <>
-      <div>AdminQuickRound</div>
-    </>
+    <div className="bg-gradient-to-r from-violet-500 to-sky-500 min-h-screen w-full">
+      <div className="flex justify-center items-center flex-col p-6">
+        <div className="text-4xl p-6">Admin quick round</div>
+        <Button addNewContent={"w-40"} onClick={openModal}>
```
Comment (MDGcode) [inline client/src/views/AdminQuickRound.tsx]: Replace the add button with a fixed + icon and add it in the corner of the screen

File: client/src/views/AdminQuickRound.tsx
Comment (MDGcode) [inline client/src/views/AdminQuickRound.tsx:1]: As a overall feedback on the design, I would change the text from black to white, and rethink the spacing and proportions of the page. Also try to use more brighter colors for the buttons to set them apart from the other components

File: client/src/components/Modal.tsx
Comment (MDGcode) [inline client/src/components/Modal.tsx:1]: Again with the proportions and the text.
Replace the close button with and X icon and stick it in the corner

File: client/src/newtork/adminQuickRound.ts [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+      "Content-Type": "application/json",
+      Authorization: `Bearer ${token}`,
+    },
+  });
+
+  return await response.json();
+};
+
+export const deleteQuestion = async (id: number) => {
+  const response = await fetch(`${url}/${id}/`, {
```
Comment (MDGcode) [inline client/src/newtork/adminQuickRound.ts]: Add /api/questions/ to every request

File: client/src/components/QuestionBox.tsx (hunk 24 above)
Comment (bogdanishere) [reply client/src/components/QuestionBox.tsx]: If we end up with 100 or more questions, we’ll need to change the endpoint logic because loading all 100 at once would take too long. To fix this, we should add pagination, so we can show a smaller number of items per page.

File: client/src/views/PageNotFetch.tsx
Comment (bogdanishere) [reply client/src/views/PageNotFetch.tsx:1]: yes, that way we can see if there is an issue with the requests.

File: client/src/components/EditQuestionForm.tsx (hunk 13 above)
Comment (bogdanishere) [reply client/src/components/EditQuestionForm.tsx:52]: I added validation directly in the input field, so the user can't enter incorrect values.
